Renovate `putter` and remove some data races

mhagger opened the pull request:
This is the next (last?) installment of refactorings to the `putter` code. It changes the code to use some more modern approaches, like contexts, `errgroup.Group`, and an `io.Pipe` to coordinate the goroutines and terminate them gracefully, hopefully without any more data races. It's based on the `s3-client-for-putter` branch (#17) so I'll leave it in draft mode until that PR is reviewed and merged, then probably rebase it.

It can be read commit-by-commit, but that third-to-last commit is a doozy so it might be just as simple to review the whole thing at once (relative to `s3-client-for-putter`). Happily, that other PR removed a lot of verbiage from this file, so I think it's fairly grokable. There's a big comment explaining how things work.

/cc @ccstolley, @carlosmn

mhagger committed:
putter.Close(): close the channel before waiting on `wg`

The only method that writes into the channel is `flush()`, and that is
only called by `Write()` and `Close()`. Therefore, after `Close()` has
made its call to `flush()`, there is no reason for it not to close the
channel immediately. This makes the next step easier.

mhagger committed:
newPutter(): initialize the `putter` instance more directly

mhagger committed:
putter.retryPutPart(): if there's an error, return it

This is the first step in solving a data race involving `p.err`.

mhagger committed:
putter.retryPutPart(): inline method

It wasn't doing much anymore.

mhagger committed:
putter.wg: change what the `WaitGroup` counts

Previously, the wait group was incremented when `flush()` was about to
put a part into `putter.ch`, and decremented when `retryPutPart()`
exited after having dispensed with one part. This is unconventional.
Instead, let's count the number of workers that are still alive. Once
the channel is closed, the workers will all exit, which is an event
that we can wait for instead. This will make it easier to allow early
cancellation of the workers if there is an error without having to
worry about the part bookkeeping going awry.

mhagger committed:
putter: add a context with the possibility of cancellation

And ensure that it gets canceled by `Close()`.

mhagger committed:
putter.flush(): if context is canceled, stop sending parts

mhagger committed:
putter: rationalize the management of goroutines

The old code had a few races, especially in the way that it accessed
`p.err` from multiple goroutines without synchronization. It also had
to do quite a bit of bookkeeping to split `Write()` data into
buffers (and crashy if more than one `Write()` call is made
concurrently).

So change the dataflow and how goroutines are managed:

* Get rid of `putter.err`. Instead, use an `errgroup.Group` to start
  the worker goroutines and collect their errors.

* Instead of `Write()` splitting the data into buffers by hand, use an
  `io.Pipe`. `Write()` writes into its write end, and a new worker,
  `queueParts()` reads from its read end using `io.ReadFull()` and
  passes the parts along to the workers.

* On errors, cancel the context and close the write end of the pipe.
  This (relatively) promptly unblocks all of the workers and allows
  them to clean up after themselves.

* Add a long comment how this all works together.

This change required test `TestPutterAfterError` to be temporarily
disabled, because it relied on being able to write directly to
`putter.err` (which, of course, was also racy).

mhagger committed:
putter.ctx: turn field into a local variable

mhagger committed:
putter.ch: turn field into a local variable

mhagger commented:
I just rebased this branch onto the new `master`. There were no conflicts and no code changes.

carlosmn commented on putter.go:
```diff
@@ -65,80 +113,132 @@ type putter struct {
-	p.c.Concurrency = max(c.Concurrency, 1)
-	p.c.NTry = max(c.NTry, 1)
-	p.bufsz = max64(minPartSize, c.PartSize)
+	ctx, cancel := context.WithCancel(context.TODO())
```
`context.TODO()` implies work to be done to determine the right context to use, so this should likely be 
```suggestion
	ctx, cancel := context.WithCancel(context.Background())
```

mhagger replied:
I was thinking that it should be possible for the caller to pass a context in here, but I guess that doesn't really fit with the `io.Writer` paradigm. So I've committed your suggestion.

(On the other hand, there should perhaps be a public `Abort()` command to be used to call the whole thing off. But that doesn't need to be part of this PR.)

carlosmn reviewed (approved):
Looks like a pretty straightforward conversion (most of the commits anyway). Just a comment about the use of `context.TODO()`.

mhagger committed:
Change how `putter` context is initialized

Co-authored-by: Carlos Martín Nieto <carlosmn@github.com>

mhagger committed:
Pass `url.URL`s by reference

It's usually a good policy to imitate the standard library.

mhagger committed:
Config.safeCopy(): new helper method

mhagger committed:
s3client.Client: bake the md5 URL into the struct

This means that `putter` doesn't need to have a `*Bucket` member
anymore, and will make it easier to mock the client interface.

mhagger committed:
putter.url: remove field

We don't need this field anymore, because all of the URLs that we need
are baked into the `s3client.Client`.

mhagger committed:
newPutter(): take the `s3Putter` as an argument

Instead of constructing an `s3client.Client`, change `newPutter()` to
take an `s3Putter` as an argument. This will make it easier to fake.

mhagger committed:
putter.Write(): if a worker has failed, return that error

If a worker fails, unblock and inform any writers by making their
`Write()` calls fail with the error that occurred during the failed
upload.

mhagger committed:
putter: add some tests of error handling

This effectively replaces the test that had to be disabled earlier in
this patch series, and also tests some other error handling scenarios.

mhagger commented:
I merged @carlosmn's suggested change, added some testability features and added some more tests, including the one that I had to disable earlier. I also tweaked `Write()` to return the error of a failing upload attempt. Sorry for the churn but I think this is good to go now.
